libyul/optimiser/LoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/LoopInvariantCodeMotion.h>
+
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/SSAValueTracker.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+#include <boost/range/adaptors.hpp>
+
+#include <utility>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void LoopInvariantCodeMotion::operator()(Block& _block)
+{
+	iterateReplacing(
+		_block.statements,
+		[&](Statement& _s) -> std::optional<vector<Statement>>
+		{
+			visit(_s);
+			if (_s.type() == typeid(ForLoop))
+			{
+				ForLoop& loop = boost::get<ForLoop>(_s);
+				return { rewriteLoop(loop) };
+			}
+			return {};
+		}
+	);
+}
+
+vector<Statement> LoopInvariantCodeMotion::rewriteLoop(ForLoop& _for)
+{
+	assertThrow(_for.pre.statements.empty(), OptimizerException, "");
+	vector<Statement> stmts;
+	// A declaration can be promoted iff
+	// 1. Its LHS is a SSA variable
+	// 2. Its RHS does not references variables declared in current scope
+	// 3. Its RHS is movable
+	auto canBePromoted = [&](Statement& _s, set<YulString> const& _varsDefinedInCurrentScope) -> bool {
+		if (_s.type() == typeid(VariableDeclaration))
+		{
+			VariableDeclaration& varDecl = boost::get<VariableDeclaration>(_s);
+			for (auto const& var: varDecl.variables)
+				if (!m_SSAVariables.count(var.name))
+					return false;
+			for (YulString varName: ReferencesCounter::countReferences(*varDecl.value) | boost::adaptors::map_keys)
+				if (_varsDefinedInCurrentScope.count(varName))
+					return false;
+			return SideEffectsCollector{m_dialect, *varDecl.value, m_functionSideEffects}.movable();
+		}
+		else
+			return false;
+	};
+	auto promoteSSADecl = [&](Block& _block) {
+		set<YulString> varsDefinedInScope;
+		iterateReplacing(
+			_block.statements,
+			[&](Statement& _s) -> std::optional<vector<Statement>>
+			{
+				if (canBePromoted(_s, varsDefinedInScope))
+				{
+					stmts.emplace_back(std::move(_s));
+					// Do not add the variables declared here to varsDefinedInScope because we are moving them.
+					return std::optional<vector<Statement>>{std::in_place};
+				}
+				else if (_s.type() == typeid(VariableDeclaration))
+				{
+					VariableDeclaration& varDecl = boost::get<VariableDeclaration>(_s);
```

Comment:
Local variable needed?

Reply:
Just for readability. I can combine it with line below if that is better.

Reply:
I think it is more needed here than the loop above :)